Remove protocol restriction for SRV records in cloudflare_dns module. Fixes #36708

DennisGlindhart opened the pull request:
##### SUMMARY
Remove protocol restriction for SRV records in cloudflare_dns module. Fixes #36708

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
cloudflare_dns

##### ANSIBLE VERSION
```
ansible 2.4.3.0
  python version = 2.7.14 (default, Feb 17 2018, 10:42:17) [GCC 7.3.1 20180130 (Red Hat 7.3.1-2)]

```

Akasurde commented on lib/ansible/modules/net_tools/cloudflare_dns.py:
```diff
@@ -44,7 +44,7 @@
     description: Service protocol. Required for C(type=SRV)
     required: false
-    choices: [ 'tcp', 'udp' ]
+    choices: [ 'tcp', 'udp', 'tls' ]
```
Could you please mention like this in description 

```
option 'tls' is added version 2.6
```

andreaso commented:
Ok, up until I had simply assumed that _tcp_ and _udp_ were the only two allowed values for the SRV record protocol field. Yet, taking a closer look at [RFC 2782](https://tools.ietf.org/html/rfc2782) those two protocols are simply listed as _"at present the most useful values for this field"_.

Likewise CloudFlare appear to accept all kinds of values in the protocol position, not simply _tcp_, _udp_, and _tls_.

@DennisGlindhart Is there any specific and/or common use case for naming the _tls_ protocol in SRV records? That is, does it make the most sense to explicit add _tls_ as an allowed parameter value, or does it make more sense to fully remove the present restrictions?

DennisGlindhart commented:
@andreaso According to RFC 2782 all kinds of values are allowed, yes. Although, using the Cloudflare Admin-UI in the browser, it seems that only tcp, udp and tls are possible to create in CF. It might be worth investigating whether or not CF allows for other types when using the API instead.

If CF really does support arbitrary values, I think it would make sense to remove the restriction completely.

I'll try to investigate further and adjust PR accordingly.

andreaso commented:
Using the API (or more specifically, by using a locally modified cloudflare_dns module) I managed to create a SRV record with the made-up _zebra_ protocol.

On the other hand, there is a benefit in keeping a restrictive parameter list, as it makes it easier to use the module properly. Especially since it's to my knowledge is very unusual for people to need SRV records for arbitrary protocols.

Hence back to my original question, what's the use case for the _tls_ protocol in this context?

DennisGlindhart commented:
@andreaso Yep - I was also able to create some "random" protocol via the API.
My particular case for tls was setting up some Office 365/Lync
``
SRV _sip._tls.example.com > sipdir.online.lync.com
``
(Example here: https://support.microsoft.com/en-us/help/2566790/troubleshooting-skype-for-business-online-dns-configuration-issues-in)

So TL;DR: Cloudflare API allows us to follow the RFC and use arbitrary protocol. Although tcp, udp and tls are probably enough for 99% of users, there are valid scenarios for other protocols as well. (Maybe we will soon see more protocols used (i.e SCTP or DCCP) when NAT is more or less going away with IPv6 currently preventing use of other protocols than TCP and UDP)

I don't think removing the restriction makes the module that much harder to use, and following the standard  is more important when we have the possibility. We can write the "normal" values (tcp & udp) in the description. I think people using this module will mostly be copying the values from some reference (like the link above) - or alternatively - know what they are doing.

I'm thinking of just remove the restrictions then - Agree?

andreaso commented:
Thanks for helping me sort out what is what here! At this point I can see the merit with both approaches, and I'll be willing to approve whichever one you choose to move forward with in this pull request.

DennisGlindhart commented:
I have now removed the restriction altogether so we follow the RFC.

I've also "rebased" / removed my previous commits (adding tls) so this is one clean commit and changed the title of the PR to be more correct. I think this should be ready for merge now.

Thanks for all your help!

cc @andreaso

Akasurde commented on lib/ansible/modules/net_tools/cloudflare_dns.py:
```diff
@@ -42,9 +42,8 @@
     default: "1"
   proto:
-    description: Service protocol. Required for C(type=SRV)
+    description: Service protocol. Required for C(type=SRV). Common values are tcp and udp. (Before Ansible 2.6 only tcp and udp were available)
```
Missing . at end.

Akasurde replied:
Also, add this note is [CHANGELOG.md](https://github.com/ansible/ansible/blob/devel/CHANGELOG.md#minor-changes) telling end-user about this.

DennisGlindhart committed:
Remove restriction of SRV protocols

andreaso commented:
shipit

andreaso commented:
@Akasurde: Happy as well?

DennisGlindhart commented:
@Akasurde I adjusted according to your review/change requests and pushed a new commit replacing the existing one with the intention to to keep history simple, but the automated bot/checks does not seem to have picked up the change (maybe because it wasn't a new commit on top of the existing one).
TL;DR Your requested changes have been implemented, so this should AFAIK be ready for merge :)

gundalow reviewed (approved):
Assuming `udp` and `UdP` work the same. Otherwise use tolower

andreaso commented:
Yes, as a general rule DNS names are compared without considering case.

When it comes to SRV records [RFC 2782](https://tools.ietf.org/html/rfc2782) explicitly  states that _"The Proto is case insensitive."_

gundalow commented:
Thanks for confirming

Akasurde commented:
rebuild_merge

Akasurde commented:
@DennisGlindhart @andreaso @gundalow Thanks for contribution.
